### DISCUSSION TITLE
Could Tasmota support the Emporia Vue Gen 2?

### DISCUSSION THREAD

**Discussion by crlogic** [arendst/Tasmota#12749] [Ideas]
I have never used Tasmota before but know of it.  Have played with ESP32's & micropython so flashing things isn't too scary.

I just got an Emporia Vue Gen 2 with two CT's for mains just to explore it.

Opening it up it has a WROVER-B, a single SAMD09D and two HC4051's.  There are headers and labels so I don't think flashing it would be a challenge.

Sadly Emporia does not offer local MQTT and rate-limits their non-public API that HA taps into.

I would be willing to flash it if someone has a bin that might support energy monitoring.  I realize I have no software development skills and that SAMD likely does magic..

[edit] WROVER not WROOM

**Comment by sfromis** [comment]
With another MCU, the SAMD09D, it sounds like some work may be needed to explore how the device works, and possibly get it supported.  A common setup with another MCU is that the ESP8266 is merely used as a "wifi modem" with the other processor running the show, and having serial communication. While this may mean that the rest of the hardware is "taken care of", it is likely an open question how that communication works.

**Reply by crlogic** [reply]
That makes sense.  

What would be the best way to go about exploring this?

**Comment by digiblur** [comment]
> SAMD09D

Time to do some serial sniffing if that is the interface it is using.  Can you pin things out to see what is connected to what on each chip between the ESP and secondary MCU?

**Reply by crlogic** [reply]
I can take some pictures with the DSLR tomorrow since I'll have it apart to add headers ;)

**Reply by crlogic** [reply]
Progress so far;

1. Pictures taken; https://imgur.com/a/FtPPLel
2. Header pins soldered and it powered on via serial-to-USB (D3 solid green).
3. I don't think I could get it to enter programming mode using the IO0 header
4. Grounding GPIO pin manually on the ESP made it long beep and D4 now blinks green.
5. No serial output using TX/RX headers (watching putty)
6. Manually holding cables to GPIO1 & GPIO3 I see stuff in putty; will save a log once better connection made.
7. No results trying to read_flash using esptool atm, will continue to poke - suggestions welcome - complete nOOb here

**Reply by digiblur** [reply]
You might need to check the efuses https://github.com/espressif/esptool/wiki/espefuse

**Reply by crlogic** [reply]
> You might need to check the efuses https://github.com/espressif/esptool/wiki/espefuse

Headers soldered directly to GPIO1/3 and I can see that I can put it into 'download mode'.

Here is what I get from espefuse - looks like most are unburnt (0). Look ok to you?

> espefuse.py --port COM20  summary
Connecting........
Detecting chip type... ESP32
espefuse.py v3.1
EFUSE_NAME (Block)                       Description  = [Meaningful Value] [Readable/Writeable] (Hex Value)
----------------------------------------------------------------------------------------
Calibration fuses:
BLK3_PART_RESERVE (BLOCK0):              BLOCK3 partially served for ADC calibration data   = False R/W (0b0)
ADC_VREF (BLOCK0):                       Voltage reference calibration                      = 1170 R/W (0b01010)

Config fuses:
XPD_SDIO_FORCE (BLOCK0):                 Ignore MTDI pin (GPIO12) for VDD_SDIO on reset     = False R/W (0b0)
XPD_SDIO_REG (BLOCK0):                   If XPD_SDIO_FORCE, enable VDD_SDIO reg on reset    = False R/W (0b0)
XPD_SDIO_TIEH (BLOCK0):                  If XPD_SDIO_FORCE & XPD_SDIO_REG                   = 1.8V R/W (0b0)
CLK8M_FREQ (BLOCK0):                     8MHz clock freq override                           = 49 R/W (0x31)
SPI_PAD_CONFIG_CLK (BLOCK0):             Override SD_CLK pad (GPIO6/SPICLK)                 = 0 R/W (0b00000)
SPI_PAD_CONFIG_Q (BLOCK0):               Override SD_DATA_0 pad (GPIO7/SPIQ)                = 0 R/W (0b00000)
SPI_PAD_CONFIG_D (BLOCK0):               Override SD_DATA_1 pad (GPIO8/SPID)                = 0 R/W (0b00000)
SPI_PAD_CONFIG_HD (BLOCK0):              Override SD_DATA_2 pad (GPIO9/SPIHD)               = 0 R/W (0b00000)
SPI_PAD_CONFIG_CS0 (BLOCK0):             Override SD_CMD pad (GPIO11/SPICS0)                = 0 R/W (0b00000)
DISABLE_SDIO_HOST (BLOCK0):              Disable SDIO host                                  = False R/W (0b0)

Efuse fuses:
WR_DIS (BLOCK0):                         Efuse write disable mask                           = 0 R/W (0x0000)
RD_DIS (BLOCK0):                         Efuse read disable mask                            = 0 R/W (0x0)
CODING_SCHEME (BLOCK0):                  Efuse variable block length scheme
   = NONE (BLK1-3 len=256 bits) R/W (0b00)
KEY_STATUS (BLOCK0):                     Usage of efuse block 3 (reserved)                  = False R/W (0b0)

Identity fuses:
MAC (BLOCK0):                            Factory MAC Address
   = xx:xx:xx:xx:xx:xx (CRC 0xa2 OK) R/W
MAC_CRC (BLOCK0):                        CRC8 for factory MAC address                       = 162 R/W (0xa2)
CHIP_VER_REV1 (BLOCK0):                  Silicon Revision 1                                 = True R/W (0b1)
CHIP_VER_REV2 (BLOCK0):                  Silicon Revision 2                                 = False R/W (0b0)
CHIP_VERSION (BLOCK0):                   Reserved for future chip versions                  = 2 R/W (0b10)
CHIP_PACKAGE (BLOCK0):                   Chip package identifier                            = 1 R/W (0b001)
MAC_VERSION (BLOCK3):                    Version of the MAC field                           = 0 R/W (0x00)

Security fuses:
FLASH_CRYPT_CNT (BLOCK0):                Flash encryption mode counter                      = 0 R/W (0b0000000)
UART_DOWNLOAD_DIS (BLOCK0):              Disable UART download mode (ESP32 rev3 only)       = False R/W (0b0)
FLASH_CRYPT_CONFIG (BLOCK0):             Flash encryption config (key tweak bits)           = 0 R/W (0x0)
CONSOLE_DEBUG_DISABLE (BLOCK0):          Disable ROM BASIC interpreter fallback             = True R/W (0b1)
ABS_DONE_0 (BLOCK0):                     Secure boot V1 is enabled for bootloader image     = False R/W (0b0)
ABS_DONE_1 (BLOCK0):                     Secure boot V2 is enabled for bootloader image     = False R/W (0b0)
JTAG_DISABLE (BLOCK0):                   Disable JTAG                                       = False R/W (0b0)
DISABLE_DL_ENCRYPT (BLOCK0):             Disable flash encryption in UART bootloader        = False R/W (0b0)
DISABLE_DL_DECRYPT (BLOCK0):             Disable flash decryption in UART bootloader        = False R/W (0b0)
DISABLE_DL_CACHE (BLOCK0):               Disable flash cache in UART bootloader             = False R/W (0b0)
BLOCK1 (BLOCK1):                         Flash encryption key
   = 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 R/W
BLOCK2 (BLOCK2):                         Secure boot key
   = 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 R/W
BLOCK3 (BLOCK3):                         Variable Block 3
   = 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 R/W

Flash voltage (VDD_SDIO) determined by GPIO12 on reset (High for 1.8V, Low/NC for 3.3V).

**Reply by crlogic** [reply]
I was able to run an erase_flash and write_flash using esptool per the tasmota Getting Started.

```
esptool.py --port COM20 erase_flash
esptool.py --port COM20 write_flash -fs 1MB -fm dout 0x0 tasmota.bin
```

Sadly on reboot it only displays the following;

```
rst:0x10 (RTCWDT_RTC_RESET),boot:0x13 (SPI_FAST_FLASH_BOOT)
configsip: 0, SPIWP:0x10
clk_drv:0x00,q_drv:0x01,d_drv:0x00,cs0_drv:0x02,hd_drv:0x00,wp_drv:0x04
mode:DOUT, clock div:2
cmd len 52953120
ets_main.c 371
```

I was able to restore the backup successfully. Options?

**Reply by barbudor** [reply]
Did you just flashed tasmota.bin (esp8266) on a ESP32 ?
Check the doc page dedicated to ESP32 to see how to flash

**Reply by crlogic** [reply]
> Did you just flashed tasmota.bin (esp8266) on a ESP32 ?

Oh you bet I did!  

Never even realized there were different binaries..   ..trying tasmota32.bin now :))

**Reply by digiblur** [reply]
Nice! Did you take a full backup of all of the chip?

**Reply by crlogic** [reply]
> Nice! Did you take a full backup of all of the chip?

Full?  errr, maybe..

I have a 4MB backup that I have flashed back and forth about a half dozen times before realizing this device has 4BM SPI RAM and 8MB SPI Flash..

..so it works to restore; crossing fingers that the last 4-8MB weren't needed!

**Reply by crlogic** [reply]
> > SAMD09D
> 
> Time to do some serial sniffing if that is the interface it is using. Can you pin things out to see what is connected to what on each chip between the ESP and secondary MCU?

After inspecting both top and bottom layers of the PCB aligned in pixel-layers, the SAMD appears to be directly underneath the ESP32 which makes throughholes a tough thing to track :/

**Reply by crlogic** [reply]
Found two plain text certificates and one plain text RSA private key in the backup :o

A few server addresses, ports and mqtt topics.  Although watching the serial output of the stock firmware offers up more topics.

**Comment by barbudor** [comment]
Beware that for ESP32, the flashing procedure is slightly more complex and requires more than just 1 binary. Read the doc please

**Reply by crlogic** [reply]
I RTFM'd and its alive!!!!

[edit] seriously, thank you.  I bumble my way through things the hard way some times..

**Comment by crlogic** [comment]
Now that Tasmota32 v9.5.0 is up and running on the WROVER and everything is installed back in the panel we can start poking!

Using the Berry console I see nothing on the I2C busses
```
wire2.scan()
[]
wire1.scan()
[]
```

Does that mean we want to assume the ESP32 and the SAMD are connected via SPI or UART?  Honest question, I know so very little about these things.  Need to learn more how to use it..

Direction welcome as I can't make heads or tails of https://tasmota.github.io/docs/Commands/#serial-bridge to poke for data..

**Comment by crlogic** [comment]
If anyone can point me to an example/guide that explains better than the docs how to actually uart.reading() that would be great.  As far as I can see, the official docs say what gpio does what and not how to pull a string from a serial/uart.

First day in Tasmota is very confusing..

**Reply by sfromis** [reply]
At least until now, you do not have direct serial features in Berry. To make basic use of serial bridge, after you configure pins, baud rate and SerialConfig, you can do a send to turn on reception, including selecting (via the format in the SerialSend) if you get data in hex or chars. Received data will appear in SerialReceived MQTT messages. Until you make the first send, Tasmota will not be receiving data.

**Reply by crlogic** [reply]
> At least until now, you do not have direct serial features in Berry. To make basic use of serial bridge, after you configure pins, baud rate and SerialConfig, you can do a send to turn on reception, including selecting (via the format in the SerialSend) if you get data in hex or chars. Received data will appear in SerialReceived MQTT messages. Until you make the first send, Tasmota will not be receiving data.

Few questions about this;

1. configuring pins is using the GUI options? Eg. Configuration --> Configure Modules --> GPIO1 Serial TX
2. I am able to query/confirm Baudrate/SerialConfig via MQTT correctly.  Not quite clear if the SerialSend<x> value should be raw or some json format.  Have successfully sent raw value and MQTT returned a {"SerialSend":"Done"} so I am assuming that works?   Without any other data in return am I safe to assume it is not those pins and I have to try others?

**Reply by crlogic** [reply]
What might I expect to see returned when serial RX is achieved?

```
03:27:50.853 MQT: stat/tasmota_A#####/RESULT = {"GPIO0":{"0":"None"},"GPIO1":{"0":"None"},"GPIO2":{"0":"None"},"GPIO3":{"0":"None"},"GPIO4":{"1824":"SerBr Rx"},"GPIO5":{"1792":"SerBr Tx"},"GPIO9":{"0":"None"},"GPIO10":{"0":"None"},"GPIO12":{"0":"None"},"GPIO13":{"0":"None"},"GPIO14":{"0":"None"},"GPIO15":{"0":"None"},"GPIO16":{"0":"None"},"GPIO17":{"0":"None"},"GPIO18":{"0":"None"},"GPIO19":{"0":"None"},"GPIO21":{"0":"None"},"GPIO22":{"0":"None"},"GPIO23":{"0":"None"},"GPIO25":{"0":"None"},"GPIO26":{"0":"None"},"GPIO27":{"0":"None"},"GPIO32":{"0":"None"},"GPIO33":{"0":"None"},"GPIO34":{"0":"None"},"GPIO35":{"0":"None"},"GPIO36":{"0":"None"},"GPIO39":{"0":"None"}}
03:32:03.351 MQT: stat/tasmota_A#####/RESULT = {"SSerialSend":"Done"}
03:32:03.543 MQT: tele/tasmota_A#####/RESULT = {"SSerialReceived":"1\n"}
```

So SSerialReceived seen but where is the data!?

**Reply by barbudor** [reply]
Here in the message payload : a "1" and a "\n"

**Comment by crlogic** [comment]
It seemed odd since it was just repeating the SSerialSend integer.

**Reply by sfromis** [reply]
Getting same payload back may simply be the other MCU echoing anything it gets (very common with full duplex operation), and if it does not recognize what you sent, it has nothing further to say. It might also be your tx/rx pins being connected, with you thus effectively running a loopback test. Do note that baud rate and serial config will normally be important to get right.

What format you send depends entirely on the expectations of the other MCU.

**Comment by crlogic** [comment]
I suppose I did get myself into a guessing game..  ..thx!

**Comment by chris153002** [comment]
FYI If you are still working on this.

https://flaviutamas.com/2021/reversing-emporia-vue-2

He has figured out the I2C pins (Connenction to the SAMD MCU).

**Comment by crlogic** [comment]
Wow there has been great progress. Will definitely go the ESPHome route.

**Reply by TommySharpNZ** [reply]
Did you get ESPHome running on your Emporia okay? I'm having quite a few issues with mine in that I find it quite unreliable and reboots often and loses daily counters...

**Comment by TommySharpNZ** [comment]
I have ESPHome running on mine for a week now :-)
But to be honest I'd really love to have Tasmota on it instead.... Maybe this work can somehow help get Tasmota onto it?

**Comment by hemangjoshi37a** [comment]
Hey brother. Can you please checkout my discussion #17815 please.

**Reply by TommySharpNZ** [reply]
I kindly ask that you don't spam our discussion. Your post is off-topic and is disrupting the flow of conversation. We value constructive dialogue and welcome diverse perspectives, but spamming is not productive or respectful of others' time. Please refrain from posting irrelevant content in the future.

**Reply by hemangjoshi37a** [reply]
First of all I have provided solution to this topic which discusses the author that MQTT is not available in vue devices, which can be very much helpful to the people who are in this discussion. 
And second of all who are you to decide what is spam and what is not, may be you are a spam who does not want me to spread awareness that about a great app that is available in the PlayStore.

Bro instead bashing anyone please do something productive in your life for God sake.

**Comment by MichaelWoodc** [comment]
Guys I am on the emporia vue v3 but trying to do the same thing. I am curious how much headway you guya may have made on this.  I am trying esphome and not having a seamless experience.  I prefer tasmota

**Reply by swampdogmash** [reply]
Hi all. I'm really keen to see if anyone has had any luck with the v3 now? Just looking at monitoring my power (with over 60 tamsota flashed devices around my house) and would love to keep the OS the same on a vue 3!

**Comment by MichaelWoodc** [comment]
I was not able to get tasmota to work.

I was able to make esp home work after a while.  Never circled back around
to tasmota.
